linux/hid.c

```diff
@@ -672,28 +891,38 @@ hid_device * HID_API_EXPORT hid_open_path(const char *path)
 		/* Set device error to none */
 		register_device_error(dev, NULL);
 
+		/*
+		 * Construct <sysfs_path>/device/report_descriptor using udev
+		 * First we need to make sure the incoming path is the file path (and not a symlink).
+		 * From there we'll take the final /dev/hidraw* and take the basename to get hidraw*.
+		 * With hidraw* we can use udev to locate the sysfs path to the hidraw device.
+		 * And then finally locate the report_descriptor file.
+		 */
+		char *fullpath = realpath(path, NULL);
+		char *sysname = basename(fullpath);
+		struct udev *udev;
+		struct udev_device *udev_dev;
+		udev = udev_new();
+		udev_dev = udev_device_new_from_subsystem_sysname(udev, "hidraw", sysname);
+		const char *sysfs_path = udev_device_get_syspath(udev_dev);
+		char *rpt_path = (char *)calloc(1, strlen(sysfs_path) + 25);
+		sprintf(rpt_path, "%s/device/report_descriptor", sysfs_path);
+		udev_device_unref(udev_dev);
+		udev_unref(udev);
+
 		/* Get the report descriptor */
-		int res, desc_size = 0;
+		int res;
 		struct hidraw_report_descriptor rpt_desc;
 
-		memset(&rpt_desc, 0x0, sizeof(rpt_desc));
-
-		/* Get Report Descriptor Size */
-		res = ioctl(dev->device_handle, HIDIOCGRDESCSIZE, &desc_size);
-		if (res < 0)
-			register_device_error_format(dev, "ioctl (GRDESCSIZE): %s", strerror(errno));
-
-		/* Get Report Descriptor */
-		rpt_desc.size = desc_size;
-		res = ioctl(dev->device_handle, HIDIOCGRDESC, &rpt_desc);
-		if (res < 0) {
-			register_device_error_format(dev, "ioctl (GRDESC): %s", strerror(errno));
-		} else {
+		res = get_hid_report_descriptor(rpt_path, &rpt_desc);
```

Comment:
Hmm, I agree that a bit more string manipulations than than one would wan't to see.
And yes, based on my experience, using `realpath` is a good call.

Since at this point we already have a device handle, I suggest to keep the `ioctl` approach for this scenario: it looks a bit cleaner, and less error prone (as in case with `25`, in my other comment).

Reply:
Agreed.